src/StorageRegistry.sol

```diff
@@ -0,0 +1,321 @@
+
+    /**
+     * @dev Chainlink L2 sequencer uptime feed.
+     */
+    AggregatorV3Interface public immutable uptimeFeed;
+
+    /*//////////////////////////////////////////////////////////////
+                                 STORAGE
+    //////////////////////////////////////////////////////////////*/
+
+    /**
+     * @dev Block timestamp at which this contract will no longer accept storage rent payments. Changeable by owner.
+     */
+    uint256 public deprecationTimestamp;
+
+    /**
+     * @dev Price per storage unit in USD. Fixed point value with 8 decimals, e.g. 5e8 = $5 USD. Changeable by owner.
+     */
+    uint256 public usdUnitPrice;
+
+    /**
+     * @dev Total capacity of storage units. Changeable by owner.
+     */
+    uint256 public maxUnits;
+
+    /**
+     * @dev Total number of storage units that have been rented.
+     */
+    uint256 public rentedUnits;
+
+    /*//////////////////////////////////////////////////////////////
+                               CONSTRUCTOR
+    //////////////////////////////////////////////////////////////*/
+
+    /**
+     * @notice Set the price feed, uptime feed, and initial parameters.
+     *
+     * @param _priceFeed                Chainlink ETH/USD price feed.
+     * @param _uptimeFeed               Chainlink L2 sequencer uptime feed.
+     * @param _initialDeprecationPeriod Initial deprecation period in seconds.
+     * @param _initialUsdUnitPrice      Initial unit price in USD. Fixed point value with 8 decimals.
+     * @param _initialMaxUnits          Initial maximum capacity in storage units.
+     */
+    constructor(
+        AggregatorV3Interface _priceFeed,
+        AggregatorV3Interface _uptimeFeed,
+        uint256 _initialDeprecationPeriod,
+        uint256 _initialUsdUnitPrice,
+        uint256 _initialMaxUnits
+    ) Ownable2Step() {
+        priceFeed = _priceFeed;
+        uptimeFeed = _uptimeFeed;
+        deprecationTimestamp = block.timestamp + _initialDeprecationPeriod;
+        usdUnitPrice = _initialUsdUnitPrice;
+        maxUnits = _initialMaxUnits;
+    }
+
+    /*//////////////////////////////////////////////////////////////
+                               MODIFIERS
+    //////////////////////////////////////////////////////////////*/
+
+    modifier whenNotDeprecated() {
+        if (block.timestamp >= deprecationTimestamp) revert ContractDeprecated();
+        _;
+    }
+
+    /*//////////////////////////////////////////////////////////////
+                        STORAGE RENTAL LOGIC
+    //////////////////////////////////////////////////////////////*/
+
+    /**
+     * @notice Rent storage for a given fid. The caller must provide an exact payment amount.
+     *
+     * @param fid   The fid that will receive the storage allocation.
+     * @param units Number of storage units to rent.
+     */
+    function rent(uint256 fid, uint256 units) external payable whenNotDeprecated {
+        if (msg.value != price(units)) revert InvalidPayment();
```

Comment:
Should we accept overpayments and return the excess funds? The concern is that we might be reverting transactions and frustrating users who used a slightly higher exchange rate